src/stan/services/util/initialize.hpp

```diff
@@ -13,202 +13,211 @@
+ * parameters once to see if it's valid.
+ *
+ * When at least some of the initialization is random, it will
+ * randomly initialize until it finds a set of unconstrained
+ * parameters that are valid or it hits <code>MAX_INIT_TRIES =
+ * 100</code> (hard-coded).
+ *
+ * Valid initialization is defined as a finite, non-NaN value
+ * for the evaluation of the log probability and all its
+ * gradients.
+ *
+ * @param[in] model the model
+ * @param[in] init a var_context with initial values
+ * @param[in,out] rng random number generator
+ * @param[in] init_radius the radius for generating random values.
+ *   A value of 0 indicates that the unconstrained parameters (not
+ *   provided by init) should be initialized with 0.
+ * @param[in] print_timing indicates whether a timing message should
+ *   be printed to the logger
+ * @param[in,out] logger logger for messages
+ * @param[in,out] init_writer init writer (on the unconstrained scale)
+ * @throws exception passed through from the model if the model has a
+ *   fatal error (not a std::domain_error)
+ * @throws std::domain_error if the model can not be initialized and
+ *   the model does not have a fatal error (only allows for
+ *   std::domain_error)
+ * @return valid unconstrained parameters for the model
+ */
+template <class Model, class RNG>
+std::vector<double> initialize(Model& model,
+                               stan::io::var_context& init,
+                               RNG& rng,
+                               double init_radius,
+                               bool print_timing,
+                               stan::callbacks::logger& logger,
+                               stan::callbacks::writer&
+                               init_writer) {
+  std::vector<double> unconstrained;
+  std::vector<int> disc_vector;
 
-        bool is_fully_initialized = true;
-        bool any_initialized = false;
-        std::vector<std::string> param_names;
-        model.get_param_names(param_names);
-        for (size_t n = 0; n < param_names.size(); n++) {
-          is_fully_initialized &= init.contains_r(param_names[n]);
-          any_initialized |= init.contains_r(param_names[n]);
-        }
-
-        bool init_zero = init_radius <= std::numeric_limits<double>::min();
+  bool is_fully_initialized = true;
+  bool any_initialized = false;
+  std::vector<std::string> param_names;
+  model.get_param_names(param_names);
+  for (size_t n = 0; n < param_names.size(); n++) {
+    is_fully_initialized &= init.contains_r(param_names[n]);
+    any_initialized |= init.contains_r(param_names[n]);
+  }
 
-        int MAX_INIT_TRIES = is_fully_initialized || init_zero ? 1 : 100;
-        int num_init_tries = 0;
-        for (; num_init_tries < MAX_INIT_TRIES; num_init_tries++) {
-          stan::io::random_var_context
-            random_context(model, rng, init_radius, init_zero);
+  bool init_zero = init_radius <= std::numeric_limits<double>::min();
```

Comment:
[required]

This needs documentation.  The name `init_zero` also needs to change as it's not testing for zero, it's testing for smaller than smallest normalized floating point number.  That means there are values other than 0 that this passes for.

[optional]

This could just change to `== 0` and the only thing you'd miss is unnormalized small numbers and it wouldn't need any doc or name changes.

Reply:
I'll leave an inline comment. There's also a function-level comment:
>  * When the <code>var_context</code> provides all variables or
>  * the init_radius is 0, this function will only evaluate the
>  * log probability of the model with the unconstrained
>  * parameters once to see if it's valid.
> ...
> * @param[in] init_radius the radius for generating random values.
> *   A value of 0 indicates that the unconstrained parameters (not
>  *   provided by init) should be initialized with 0.

I'm sure I could doc it better; what documentation do I need to leave behind so this is clear?

Reply:
I changed the name to `is_initialized_with_zero` and I think there isn't much I can do with an inline comment now. Just let me know if there's anything else that I can add to the doc.